gusto/linear_solvers.py

```diff
@@ -229,6 +234,327 @@ def solve(self):
+
+        # X = A.inv * (X_r - K * l),
+        # 0 = K.T * X = -(K.T * A.inv * K) * l + K.T * A.inv * X_r,
+        # so (K.T * A.inv * K) * l = K.T * A.inv * X_r
+        # is the reduced equation for the Lagrange multipliers.
+        # Right-hand side expression: (Forward substitution)
+        Rexp = K.T * A.inv * X_r
+        self.R = Function(Vtrace)
+
+        # We need to rebuild R everytime data changes
+        self._assemble_Rexp = create_assembly_callable(Rexp, tensor=self.R)
+
+        # Schur complement operator:
+        Smatexp = K.T * A.inv * K
+        S = assemble(Smatexp)
+        S.force_evaluation()
+
+        # Set up the Linear solver for the system of Lagrange multipliers
+        self.lSolver = LinearSolver(S, solver_parameters=self.solver_parameters,
+                                    options_prefix='lambda_solve')
+
+        # Result function for the multiplier solution
+        self.lambdar = Function(Vtrace)
+
+        # Place to put result of u rho reconstruction
+        self.urho = Function(M)
+
+        # Reconstruction of broken u and rho
+        u_, rho_ = self.urho.split()
+
+        # Split operators for two-stage reconstruction
+        A00 = A.block((0, 0))
+        A01 = A.block((0, 1))
+        A10 = A.block((1, 0))
+        A11 = A.block((1, 1))
+        K0 = K.block((0, 0))
+        Ru = X_r.block((0,))
+        Rrho = X_r.block((1,))
+        lambda_vec = AssembledVector(self.lambdar)
+
+        # rho reconstruction
+        Srho = A11 - A10 * A00.inv * A01
+        rho_expr = Srho.inv * (Rrho - A10 * A00.inv * (Ru - K0 * lambda_vec))
+        self._assemble_rho = create_assembly_callable(rho_expr, tensor=rho_)
+
+        # "broken" u reconstruction
+        rho_vec = AssembledVector(rho_)
+        u_expr = A00.inv * (Ru - A01 * rho_vec - K0 * lambda_vec)
+        self._assemble_u = create_assembly_callable(u_expr, tensor=u_)
+
+        # Project broken u into the HDiv space using facet averaging.
+        # Weight function counting the dofs of the HDiv element:
+        shapes = (Vu.finat_element.space_dimension(), np.prod(Vu.shape))
+
+        weight_kernel = """
+        for (int i=0; i<%d; ++i) {
+        for (int j=0; j<%d; ++j) {
+        w[i][j] += 1.0;
+        }}""" % shapes
+
+        self._weight = Function(Vu)
+        par_loop(weight_kernel, dx, {"w": (self._weight, INC)})
+
+        # Averaging kernel
+        self._average_kernel = """
+        for (int i=0; i<%d; ++i) {
+        for (int j=0; j<%d; ++j) {
+        vec_out[i][j] += vec_in[i][j]/w[i][j];
+        }}""" % shapes
+
+        # HDiv-conforming velocity
+        self.u_hdiv = Function(Vu)
+
+        # Reconstruction of theta
+        theta = TrialFunction(Vtheta)
+        gamma = TestFunction(Vtheta)
+
+        self.theta = Function(Vtheta)
+        theta_eqn = gamma*(theta - theta_in +
+                           dot(k, self.u_hdiv)*dot(k, grad(thetabar))*beta)*dx
```

Comment:
Hmm, I've realised that this isn't right as `thetabar` has been scaled by `1/(1+water_t)` while `theta_in` has not (definitely my bad, so it'll also be wrong in the `CompressibleSolver`)

Reply:
So basically we just need something like `theta_in = 1/(1+water_t) * theta_in`?

Reply:
I can take the opportunity to correct this in this PR since it seems to be a small update

Reply:
Or if you want, you can check out this branch and push your commit to it